R/derive_advs_params.R

```diff
@@ -0,0 +1,633 @@
+#'   which is required to derive the new parameter is measured only once. For example,
+#'   if Height to Weight Ratio should be derived and height is measured only once while
+#'   Weight is measured at each visit. Height could be specified in the `numerator_code`
+#'   argument and `constant_numerator` is to be set to `TRUE`.
+#'
+#'   *Permitted Values:* logical scalar
+#'
+#' @param constant_denominator Is denominator parameter constant?
+#'
+#'   It is expected that the parameter code (PARAMCD) specified in `numerator_code`
+#'   which is required to derive the new parameter is measured only once. For example,
+#'   if Waist to Height Ratio should be derived and height is measured only once
+#'   while Waist Circumference is measured at each visit. Height could be specified in
+#'   the `denominator_code` argument and `constant_denominator` is to be set to `TRUE`.
+#'
+#'   *Permitted Values:* logical scalar
+#'
+#' @param constant_by_vars By variables for when numerator and/or denominator is constant
+#'
+#'   When numerator and/or denominator is constant, the parameters (measured only once) are merged
+#'   to the other parameters using the specified variables.
+#'
+#'   If numerator and/or denominator is constant (e.g. only measured once at screening or baseline)
+#'   then use `constant_by_vars` to select the subject-level variable to merge on (e.g. `USUBJID`).
+#'   This will produce a generic Ratio parameter at all visits where numerator and/or denominator
+#'   is measured. Otherwise it will only be calculated at visits with both numerator and denominator
+#'   parameters collected.
+#'
+#'   *Permitted Values*: list of variables created by `exprs()`, e.g. `exprs(USUBJID, VISIT)`
+#'
+#' @param get_unit_expr An expression providing the unit of the parameter
+#'
+#'   The result is used to check the units of the input parameters. If the units are not consistent
+#'   within each parameter, an error will be thrown.
+#'
+#'   Additionally, if the input parameters are measured in different units but are mutually
+#'   convertible (e.g., centimeters for one parameter and inches for another), an automatic
+#'   conversion will be performed in order to uniform the values before calculating the ratio.
+#'
+#'   **Note:** Conversion factors come from unit definitions as per CDISC standards.
+#'   ```{r, echo = FALSE, comment = "", results = "asis"}
+#'   get_conv_factors_all()[["length"]] %>%
+#'     discard_at("cm") %>%
+#'     str_glue_data("<br/>*{names(.)}* is defined as {.} cm")
+#'   ```
+#'
+#'   *Permitted Values:* A variable of the input dataset or a function call
+#'
+#' @inheritParams admiral::derive_param_computed
+#'
+#' @details
+#' The analysis value of the new parameter is derived as
+#' \deqn{RATIO = \frac{NUMERATOR}{DENOMINATOR}}
+#'
+#'
+#' @return The input dataset with the new parameter added. Note, a variable will only
+#'         be populated in the new parameter rows if it is specified in `by_vars`.
```

Comment:
What does "Note, a variable will only be populated in the new parameter rows if it is specified in `by_vars`."

What does this mean? I did not think that `by_vars` had anything to do with `set_values_to`?

Reply:
This was also taken from admiral. It means that if, for instance, you have also `ACTARM` variable that you wanna keep for derived parameters, you have to specify it in `by_var` as well. Otherwise it won't get populated for new records.

```R
advs <- tribble(
  ~USUBJID, ~ACTARM, ~PARAMCD, ~PARAM, ~AVAL, ~AVALU, ~VISIT,
  "01-101-1001", "ARM 1", "HIPCIR", "Hip Circumference (cm)",   125, "cm", "SCREENING",
  "01-101-1001", "ARM 1", "HIPCIR", "Hip Circumference (cm)",   124, "cm", "WEEK 2",
  "01-101-1001", "ARM 1", "HIPCIR", "Hip Circumference (cm)",   123, "cm", "WEEK 3",
  "01-101-1001", "ARM 1", "WSTCIR", "Waist Circumference (in)", 43.31, "in", "SCREENING",
  "01-101-1001", "ARM 1", "WSTCIR", "Waist Circumference (in)", 42.52, "in", "WEEK 2",
  "01-101-1001", "ARM 1", "WSTCIR", "Waist Circumference (in)", 42.13, "in", "WEEK 3",
  "01-101-1002", "ARM 2", "HIPCIR", "Hip Circumference (cm)",   135, "cm", "SCREENING",
  "01-101-1002", "ARM 2", "HIPCIR", "Hip Circumference (cm)",   133, "cm", "WEEK 2",
  "01-101-1002", "ARM 2", "HIPCIR", "Hip Circumference (cm)",   132, "cm", "WEEK 3",
  "01-101-1002", "ARM 2", "WSTCIR", "Waist Circumference (in)", 47.24, "in", "SCREENING",
  "01-101-1002", "ARM 2", "WSTCIR", "Waist Circumference (in)", 46.46, "in", "WEEK 2",
  "01-101-1002", "ARM 2", "WSTCIR", "Waist Circumference (in)", 46.06, "in", "WEEK 3"
)

derive_param_waisthip(
  advs,
  by_vars = exprs(USUBJID, VISIT),   # No ACTARM here
  wstcir_code = "WSTCIR",
  hipcir_code = "HIPCIR",
  set_values_to = exprs(
    PARAMCD = "WAISTHIP",
    PARAM = "Waist to Hip Ratio"
  ),
  get_unit_expr = admiral::extract_unit(PARAM)
)
```

```
   USUBJID     ACTARM PARAMCD  PARAM                       AVAL AVALU VISIT    
   <chr>       <chr>  <chr>    <chr>                      <dbl> <chr> <chr>    
 1 01-101-1001 ARM 1  HIPCIR   Hip Circumference (cm)   125     cm    SCREENING
 2 01-101-1001 ARM 1  HIPCIR   Hip Circumference (cm)   124     cm    WEEK 2   
 3 01-101-1001 ARM 1  HIPCIR   Hip Circumference (cm)   123     cm    WEEK 3   
 4 01-101-1001 ARM 1  WSTCIR   Waist Circumference (in)  43.3   in    SCREENING
 5 01-101-1001 ARM 1  WSTCIR   Waist Circumference (in)  42.5   in    WEEK 2   
 6 01-101-1001 ARM 1  WSTCIR   Waist Circumference (in)  42.1   in    WEEK 3   
 7 01-101-1002 ARM 2  HIPCIR   Hip Circumference (cm)   135     cm    SCREENING
 8 01-101-1002 ARM 2  HIPCIR   Hip Circumference (cm)   133     cm    WEEK 2   
 9 01-101-1002 ARM 2  HIPCIR   Hip Circumference (cm)   132     cm    WEEK 3   
10 01-101-1002 ARM 2  WSTCIR   Waist Circumference (in)  47.2   in    SCREENING
11 01-101-1002 ARM 2  WSTCIR   Waist Circumference (in)  46.5   in    WEEK 2   
12 01-101-1002 ARM 2  WSTCIR   Waist Circumference (in)  46.1   in    WEEK 3   
13 01-101-1001 NA     WAISTHIP Waist-to-Hip Ratio         0.880 NA    SCREENING
14 01-101-1001 NA     WAISTHIP Waist-to-Hip Ratio         0.871 NA    WEEK 2   
15 01-101-1001 NA     WAISTHIP Waist-to-Hip Ratio         0.870 NA    WEEK 3   
16 01-101-1002 NA     WAISTHIP Waist-to-Hip Ratio         0.889 NA    SCREENING
17 01-101-1002 NA     WAISTHIP Waist-to-Hip Ratio         0.887 NA    WEEK 2   
18 01-101-1002 NA     WAISTHIP Waist-to-Hip Ratio         0.886 NA    WEEK 3   
```

Reply:
I see. I would argue that it could be made clearer, but I agree that we should be consistent with Admiral.